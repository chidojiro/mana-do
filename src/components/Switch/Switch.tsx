import React from 'react';

export interface Props {
  value?: boolean;
  onChange?: (val: boolean) => void;
}

export const Switch: React.FC<Props> = ({ value, onChange }) => {
  const [checked, setChecked] = React.useState(value);

  React.useEffect(() => {
    setChecked(value);
  }, [value]);

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setChecked(e.target.checked);
    onChange?.(e.target.checked);
  };

  return (
    <label className='switch'>
      <input type='checkbox' checked={checked} onChange={handleChange} />
      <span className='slider round'></span>
    </label>
  );
};
