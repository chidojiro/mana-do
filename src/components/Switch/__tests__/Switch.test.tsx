import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { Switch, Props } from '../Switch';

describe('Switch', () => {
  const defaultProps: Props = { onChange: jest.fn(), value: false };

  it('should render the component', () => {
    const wrapper = shallow(<Switch />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should call onChange with the correct state when the switch is clicked', () => {
    const wrapper = shallow(<Switch {...defaultProps} />);

    wrapper.find('input').simulate('change', { target: { checked: true } });
    expect(defaultProps.onChange).toBeCalledWith(true);

    wrapper.find('input').simulate('change', { target: { checked: false } });
    expect(defaultProps.onChange).toBeCalledWith(false);
  });

  it('should update state correctly with the correct state when the switch is clicked', () => {
    const wrapper = shallow(<Switch {...defaultProps} />);

    wrapper.find('input').simulate('change', { target: { checked: true } });
    expect(wrapper.find('input').prop('checked')).toBe(true);

    wrapper.find('input').simulate('change', { target: { checked: false } });
    expect(wrapper.find('input').prop('checked')).toBe(false);
  });

  it('should initially unchecked when value is false', () => {
    const wrapper = shallow(<Switch {...defaultProps} value={false} />);

    expect(wrapper.find('input').prop('checked')).toBe(false);
  });

  it('should toggle checked state properly', () => {
    const wrapper = shallow(<Switch {...defaultProps} value={false} />);

    expect(wrapper.find('input').prop('checked')).toBe(false);

    wrapper.find('input').simulate('change', { target: { checked: true } });
    expect(wrapper.find('input').prop('checked')).toBe(true);
  });
});
