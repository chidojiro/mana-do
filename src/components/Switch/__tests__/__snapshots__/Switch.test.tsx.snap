// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch should render the component 1`] = `
<label
  className="switch"
>
  <input
    onChange={[Function]}
    type="checkbox"
  />
  <span
    className="slider round"
  />
</label>
`;
